src/mongoose_rabbit_worker.erl

```diff
@@ -0,0 +1,306 @@
+%%% Internal functions
+%%%===================================================================
+
+do_handle_call({amqp_call, Method}, _From, State = #state{channel = Channel}) ->
+    case try_to_apply(amqp_channel, call, [Channel, Method]) of
+        {ok, _} = Res ->
+            {reply, Res, State};
+        Err ->
+            {FreshConn, FreshChann} = maybe_restart_rabbit_connection(State),
+            {reply, Err, State#state{connection = FreshConn,
+                                     channel = FreshChann}}
+    end.
+
+do_handle_cast({amqp_publish, Method, Payload}, State) ->
+    handle_amqp_publish(Method, Payload, State).
+
+do_handle_info({init, Opts}, State) ->
+    Host = proplists:get_value(host, Opts),
+    AMQPClientOpts = proplists:get_value(amqp_client_opts, Opts),
+    {Connection, Channel} = establish_rabbit_connection(AMQPClientOpts, Host),
+    IsConfirmEnabled = maybe_enable_confirms(Channel, Opts),
+    MaxMsgQueueLen = proplists:get_value(max_queue_len, Opts),
+    {noreply, State#state{host = Host, amqp_client_opts = AMQPClientOpts,
+                          connection = Connection, channel = Channel,
+                          confirms = IsConfirmEnabled,
+                          max_queue_len = MaxMsgQueueLen}}.
+
+-spec handle_amqp_publish(Method :: mongoose_amqp:method(),
+                          Payload :: mongoose_amqp:message(),
+                          Opts :: worker_opts()) -> {noreply, worker_opts()}.
+handle_amqp_publish(Method, Payload, Opts = #state{host = Host}) ->
+    {PublishTime, Result} =
+        timer:tc(fun publish_message_and_wait_for_confirm/3,
+                 [Method, Payload, Opts]),
+    case Result of
+        true ->
+            update_messages_published_metrics(Host, PublishTime, Payload),
+            ?DEBUG("event=rabbit_message_sent message=~1000p", [Payload]),
+            {noreply, Opts};
+        false ->
+            update_messages_failed_metrics(Host),
+            ?WARNING_MSG("event=rabbit_message_sent_failed reason=negative_ack",
+                         []),
+            {noreply, Opts};
+        {channel_exception, Error, Reason} ->
+            update_messages_failed_metrics(Host),
+            ?WARNING_MSG("event=rabbit_message_sent_failed reason=~1000p:~1000p",
+                         [Error, Reason]),
+            {FreshConn, FreshChann} = maybe_restart_rabbit_connection(Opts),
+            {noreply, Opts#state{connection = FreshConn, channel = FreshChann}};
+        timeout ->
+            update_messages_timeout_metrics(Host),
+            ?WARNING_MSG("event=rabbit_message_sent_failed reason=timeout",
+                         []),
+            {noreply, Opts}
+    end.
+
+-spec publish_message_and_wait_for_confirm(Method :: mongoose_amqp:method(),
+                                           Payload :: mongoose_amqp:message(),
+                                           worker_opts()) ->
+    boolean() | timeout | channel_exception.
```

Comment:
It's a matter of preference, but when a return value is complex (or there are complex parameter types involved) I like to write the specs according to the following pattern - it's more suitable for inserting line breaks for readability:

```erlang
-spec publish_message_and_wait_for_confirm(Method, Payload, worker_opts()) -> R
      when Method :: mongoose_amqp:method(),
           Payload :: mongoose_amqp:message(),
           R :: boolean() | timeout
              %% this line break is a stretch, but shows an example
              | channel_exception.
```